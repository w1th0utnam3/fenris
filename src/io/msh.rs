//! Support for loading Gmsh MSH 4.1 files as [`Mesh`]es.
//!
//! This module supports Gmsh MSH files with the file format version 4.1 as specified in the
//! [Gmsh documentation](https://gmsh.info/doc/texinfo/gmsh.html#MSH-file-format). All element types
//! that implement the [`MshConnectivity`] trait are supported.
//!
//! Example usage:
//! ```
//! use nalgebra::U2;
//! use fenris::connectivity::Tri3d2Connectivity;
//! use fenris::io::msh::load_msh_from_file;
//!
//! let mesh = load_msh_from_file::<f64, U2, Tri3d2Connectivity, _>(
//!     "assets/meshes/square_tri3_4.msh").unwrap();
//!
//! assert_eq!(mesh.vertices().len(), 5);
//! assert_eq!(mesh.connectivity().len(), 4);
//! ```

use crate::connectivity::{
    Hex27Connectivity, Hex8Connectivity, Quad4d2Connectivity, Quad9d2Connectivity, Tet10Connectivity, Tet4Connectivity,
    Tri3d2Connectivity, Tri3d3Connectivity, Tri6d2Connectivity,
};
use crate::mesh::Mesh;
use eyre::{eyre, Context};
use nalgebra::allocator::Allocator;
use nalgebra::{DefaultAllocator, DimName, OPoint, RealField};
use num::ToPrimitive;
use std::path::Path;

/// Loads a [`Mesh`] from a Gmsh MSH file at the given path.
pub fn load_msh_from_file<T, D, C, P: AsRef<Path>>(file_path: P) -> eyre::Result<Mesh<T, D, C>>
where
    T: RealField,
    D: DimName,
    C: MshConnectivity,
    DefaultAllocator: Allocator<T, D>,
{
    let msh_bytes = std::fs::read(file_path).wrap_err("failed to read file")?;
    load_msh_from_bytes(&msh_bytes).wrap_err("failed to load mesh from msh file")
}

/// Loads a [`Mesh`] by parsing the given bytes as a Gmsh MSH file.
pub fn load_msh_from_bytes<T, D, C>(bytes: &[u8]) -> eyre::Result<Mesh<T, D, C>>
where
    T: RealField,
    D: DimName,
    C: MshConnectivity,
    DefaultAllocator: Allocator<T, D>,
{
    let mut msh_file = mshio::parse_msh_bytes(bytes).map_err(|e| eyre!("failed to parse msh file: {}", e))?;

    let msh_nodes = msh_file
        .data
        .nodes
        .take()
        .ok_or(eyre!("MSH file does not contain nodes"))?;
    let msh_elements = msh_file
        .data
        .elements
        .take()
        .ok_or(eyre!("MSH file does not contain elements"))?;

    let mut vertices = Vec::new();
    let mut connectivity = Vec::new();

    // Ensure that at least one element block matches the target mesh connectivity
    if !msh_elements
        .element_blocks
        .iter()
        .any(|block| element_block_matches_connectivity::<C, _>(block))
    {
        return Err(eyre!(
            "MSH file does not contain an element block of the requested type (type {:?} of with reference/entity dim {})",
            C::msh_element_type(),
            C::reference_dim()
        ));
    }

    // Collect all mesh vertices
    for node_block in &msh_nodes.node_blocks {
        let block_vertices = vertices_from_node_block(node_block)?;
        vertices.extend(block_vertices);
    }

    if vertices.len()
        != msh_nodes
            .num_nodes
            .to_usize()
            .expect("failed to convert num_nodes to usize")
    {
        return Err(eyre!(
            "only {} vertices were read but msh file claims to contain {} nodes",
            vertices.len(),
            msh_nodes.num_nodes
        ));
    }

    // Collect all connectivity matching the target connectivity
    for element_block in &msh_elements.element_blocks {
        let block_connectivity = connectivity_from_element_block(element_block)?;
        connectivity.extend(block_connectivity);
    }

    Ok(Mesh::from_vertices_and_connectivity(vertices, connectivity))
}

/// Tries to convert a `mshio::NodeBlock` to a `Vec<OPoint<T, D>>`.
fn vertices_from_node_block<T, D, F, I>(node_block: &mshio::NodeBlock<u64, I, F>) -> eyre::Result<Vec<OPoint<T, D>>>
where
    T: RealField,
    D: DimName,
    F: mshio::MshFloatT,
    I: mshio::MshIntT,
    DefaultAllocator: Allocator<T, D>,
{
    // Ensure that node tags are consecutive
    if node_block.node_tags.is_some() {
        return Err(eyre!(
            "node block tags are not consecutive in msh file (sparse tags are not supported)"
        ));
    }

    // Note: The following check is commented out because it can lead to missing nodes.
    //  Probable explanation:
    //  The MSH `node_block`'s `entity_dim` does not seem to correspond to the geometrical
    //  dimension of the points. Rather it seems to correspond to the dimension of the "physical"
    //  object represented by the node block.
    //  When creating primitives in Gmsh for example, the nodes of a triangulation of a sphere are
    //  divided into node_blocks representing its equator, surface and volume and all of them are
    //  referenced by the volumetric elements.
    //  In addition, all node blocks have to be read anyway in order for the global `node_tag` indexing to
    //  work consistently.
    // TODO: Revisit this after some more experience with loading MSH files and delete this.
    /*
    if node_block
        .entity_dim
        .to_usize()
        .ok_or_else(|| eyre!("error converting node block entity dimension to usize"))?
        != D::dim()
    {
        return Err(eyre!("node block entity does not have the right dimension for this mesh"));
    }
    */

    let mut vertices = Vec::with_capacity(node_block.nodes.len());

    // Convert MSH vertices to points
    for node in &node_block.nodes {
        vertices.push(point_from_msh_node(node)?);
    }

    Ok(vertices)
}

/// Tries to convert a `mshio::ElementBlock` to a `Vec<Connectivity>`.
fn connectivity_from_element_block<C, I>(element_block: &mshio::ElementBlock<u64, I>) -> eyre::Result<Vec<C>>
where
    C: MshConnectivity,
    I: mshio::MshIntT,
{
    // Ensure that element tags are consecutive
    if element_block.element_tags.is_some() {
        return Err(eyre!(
            "element block tags are not consecutive in msh file (sparse tags are not supported)"
        ));
    }

    if !element_block_matches_connectivity::<C, _>(element_block) {
        // Just ignore blocks that don't match the requested connectivity
        return Ok(Vec::new());
    } else {
        let mut connectivity = Vec::with_capacity(element_block.elements.len());
        let requested_nodes = C::msh_element_type()
            .nodes()
            .map_err(|_| eyre!("unimplemented element type requested"))?;

        for element in &element_block.elements {
            if element.nodes.len() < requested_nodes {
                return Err(eyre!("not enough nodes to initialize connectivity"));
            }
            connectivity.push(C::try_connectivity_from_msh_element(element)?);
        }

        return Ok(connectivity);
    }
}

/// Returns whether the given element block contains elements corresponding to the specified connectivity.
fn element_block_matches_connectivity<C, I>(element_block: &mshio::ElementBlock<u64, I>) -> bool
where
    C: MshConnectivity,
    I: mshio::MshIntT,
{
    element_block.element_type == C::msh_element_type()
        && element_block
            .entity_dim
            .to_usize()
            .expect("failed to convert element block dimension to usize")
            == C::reference_dim()
}

macro_rules! f_to_t {
    ($component:expr) => {
        T::from_f64(
            $component
                .to_f64()
                .ok_or_else(|| eyre!("failed to convert coordinate to f64"))?,
        )
        .ok_or_else(|| eyre!("failed to convert node coordinate from f64 to target mesh real type"))?
    };
}

fn point_from_msh_node<T, D, F>(node: &mshio::Node<F>) -> eyre::Result<OPoint<T, D>>
where
    T: RealField,
    D: DimName,
    F: mshio::MshFloatT,
    DefaultAllocator: Allocator<T, D>,
{
    // TODO: Ensure that components i < D are zero?
    let mut point = OPoint::origin();
    point[0] = f_to_t!(node.x);
    if D::dim() > 1 {
        point[1] = f_to_t!(node.y);
    }
    if D::dim() > 2 {
        point[2] = f_to_t!(node.z);
    }

    Ok(point)
}

/// Allows conversion from `mshio::Element`s to connectivity types used in `fenris`.
pub trait MshConnectivity
where
    Self: Sized,
{
    /// Returns the MSH element type corresponding to this connectivity.
    fn msh_element_type() -> mshio::ElementType;
    /// Returns the reference dimension of this connectivity (corresponds to MSH entity dimension).
    fn reference_dim() -> usize;
    /// Tries to construct the element connectivity from the given MSH element.
    fn try_connectivity_from_msh_element(element: &mshio::Element<u64>) -> eyre::Result<Self>;
}

macro_rules! impl_msh_connectivity {
    ($connectivity:ident, $msh_type:ident, num_nodes = $num_nodes:literal) => {
        impl MshConnectivity for $connectivity {
            fn msh_element_type() -> mshio::ElementType {
                mshio::ElementType::$msh_type
            }

            fn reference_dim() -> usize {
                use crate::element::ElementConnectivity;
                <$connectivity as ElementConnectivity<f64>>::ReferenceDim::dim()
            }

            fn try_connectivity_from_msh_element(element: &mshio::Element<u64>) -> eyre::Result<Self> {
                assert_eq!(
                    element.nodes.len(),
                    $num_nodes,
                    "number of msh element nodes have to match with connectivity"
                );
                let mut nodes = [0; $num_nodes];
                for i in 0..$num_nodes {
                    nodes[i] = element.nodes[i] as usize - 1;
                }
                Ok(Self(nodes))
            }
        }
    };
}

impl_msh_connectivity!(Tri3d2Connectivity, Tri3, num_nodes = 3);
impl_msh_connectivity!(Tri3d3Connectivity, Tri3, num_nodes = 3);
impl_msh_connectivity!(Tri6d2Connectivity, Tri6, num_nodes = 6);
impl_msh_connectivity!(Quad4d2Connectivity, Qua4, num_nodes = 4);
impl_msh_connectivity!(Quad9d2Connectivity, Qua9, num_nodes = 9);
impl_msh_connectivity!(Tet4Connectivity, Tet4, num_nodes = 4);
impl_msh_connectivity!(Tet10Connectivity, Tet10, num_nodes = 10);
impl_msh_connectivity!(Hex8Connectivity, Hex8, num_nodes = 8);
impl_msh_connectivity!(Hex27Connectivity, Hex27, num_nodes = 27);

// The following connectivities do not implement ElementConnectivity yet
//impl_msh_connectivity!(Tri6d3Connectivity, Tri6, num_nodes = 6);
//impl_msh_connectivity!(Quad4d3Connectivity, Qua4, num_nodes = 4);
//impl_msh_connectivity!(Quad9d3Connectivity, Qua9, num_nodes = 9);
