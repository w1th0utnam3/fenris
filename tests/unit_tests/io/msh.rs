use crate::export_mesh_vtk;
use fenris::connectivity::{
    Hex27Connectivity, Hex8Connectivity, Quad4d2Connectivity, Quad9d2Connectivity, Tet10Connectivity, Tet4Connectivity,
    Tri3d2Connectivity, Tri3d3Connectivity, Tri6d2Connectivity,
};
use fenris::io::msh::load_msh_from_file;
use insta::assert_debug_snapshot;
use nalgebra::{U2, U3};

#[test]
fn load_msh_sphere_tet4_large() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U3, Tet4Connectivity, _>("assets/meshes/sphere_tet4_593.msh")?;

    assert_eq!(mesh.vertices().len(), 183);
    assert_eq!(mesh.connectivity().len(), 593);

    export_mesh_vtk("io_msh", "load_msh_sphere_tet4_large", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_rect_tri3d2_large() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U2, Tri3d2Connectivity, _>("assets/meshes/rectangle_tri3_110.msh")?;

    assert_eq!(mesh.vertices().len(), 70);
    assert_eq!(mesh.connectivity().len(), 110);

    export_mesh_vtk("io_msh", "load_msh_rect_tri3d2_large", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_rect_tri3d3_large() -> eyre::Result<()> {
    // Loading a 2D triangle mesh to a 3D triangle mesh should work
    let mesh = load_msh_from_file::<f64, U3, Tri3d3Connectivity, _>("assets/meshes/rectangle_tri3_110.msh")?;

    assert_eq!(mesh.vertices().len(), 70);
    assert_eq!(mesh.connectivity().len(), 110);

    export_mesh_vtk("io_msh", "load_msh_rect_tri3d3_large", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_square_quad4d2_large() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U2, Quad4d2Connectivity, _>("assets/meshes/square_quad4_79.msh")?;

    assert_eq!(mesh.vertices().len(), 96);
    assert_eq!(mesh.connectivity().len(), 79);

    export_mesh_vtk("io_msh", "load_msh_square_quad4d2_large", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_square_tri3d2() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U2, Tri3d2Connectivity, _>("assets/meshes/square_tri3_4.msh")?;

    assert_eq!(mesh.vertices().len(), 5);
    assert_eq!(mesh.connectivity().len(), 4);

    export_mesh_vtk("io_msh", "load_msh_square_tri3d2", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_square_tri6d2() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U2, Tri6d2Connectivity, _>("assets/meshes/square_tri6_4.msh")?;

    assert_eq!(mesh.vertices().len(), 13);
    assert_eq!(mesh.connectivity().len(), 4);

    export_mesh_vtk("io_msh", "load_msh_square_tri6d2", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_square_quad4d2() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U2, Quad4d2Connectivity, _>("assets/meshes/square_quad4_4.msh")?;

    assert_eq!(mesh.vertices().len(), 9);
    assert_eq!(mesh.connectivity().len(), 4);

    export_mesh_vtk("io_msh", "load_msh_square_quad4d2", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_square_quad9d2() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U2, Quad9d2Connectivity, _>("assets/meshes/square_quad9_4.msh")?;

    assert_eq!(mesh.vertices().len(), 25);
    assert_eq!(mesh.connectivity().len(), 4);

    export_mesh_vtk("io_msh", "load_msh_square_quad9d2", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_cube_tet4() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U3, Tet4Connectivity, _>("assets/meshes/cube_tet4_24.msh")?;

    assert_eq!(mesh.vertices().len(), 14);
    assert_eq!(mesh.connectivity().len(), 24);

    export_mesh_vtk("io_msh", "load_msh_cube_tet4", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_cube_tet10() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U3, Tet10Connectivity, _>("assets/meshes/cube_tet10_24.msh")?;

    assert_eq!(mesh.vertices().len(), 63);
    assert_eq!(mesh.connectivity().len(), 24);

    export_mesh_vtk("io_msh", "load_msh_cube_tet10", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_cube_hex8() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U3, Hex8Connectivity, _>("assets/meshes/cube_hex8_8.msh")?;

    assert_eq!(mesh.vertices().len(), 27);
    assert_eq!(mesh.connectivity().len(), 8);

    export_mesh_vtk("io_msh", "load_msh_cube_hex8", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}

#[test]
fn load_msh_cube_hex27() -> eyre::Result<()> {
    let mesh = load_msh_from_file::<f64, U3, Hex27Connectivity, _>("assets/meshes/cube_hex27_8.msh")?;

    assert_eq!(mesh.vertices().len(), 125);
    assert_eq!(mesh.connectivity().len(), 8);

    export_mesh_vtk("io_msh", "load_msh_cube_hex27", &mesh);
    assert_debug_snapshot!(mesh);

    Ok(())
}
