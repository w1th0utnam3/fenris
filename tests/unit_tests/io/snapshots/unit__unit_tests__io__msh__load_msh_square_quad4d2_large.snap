---
source: tests/unit_tests/io/msh.rs
expression: mesh
---
Mesh {
    vertices: [
        [
            0.0,
            0.0,
        ],
        [
            0.5,
            0.0,
        ],
        [
            0.5,
            0.5,
        ],
        [
            0.0,
            0.5,
        ],
        [
            0.0625000000000001,
            0.0,
        ],
        [
            0.1250000000000001,
            0.0,
        ],
        [
            0.1875000000000001,
            0.0,
        ],
        [
            0.2500000000000001,
            0.0,
        ],
        [
            0.3125000000000001,
            0.0,
        ],
        [
            0.3750000000000001,
            0.0,
        ],
        [
            0.4375,
            0.0,
        ],
        [
            0.5,
            0.0625000000000001,
        ],
        [
            0.5,
            0.1250000000000001,
        ],
        [
            0.5,
            0.1875000000000001,
        ],
        [
            0.5,
            0.2500000000000001,
        ],
        [
            0.5,
            0.3125000000000001,
        ],
        [
            0.5,
            0.3750000000000001,
        ],
        [
            0.5,
            0.4375,
        ],
        [
            0.4374999999999999,
            0.5,
        ],
        [
            0.3749999999999999,
            0.5,
        ],
        [
            0.3124999999999999,
            0.5,
        ],
        [
            0.2499999999999999,
            0.5,
        ],
        [
            0.1874999999999999,
            0.5,
        ],
        [
            0.1249999999999999,
            0.5,
        ],
        [
            0.0625,
            0.5,
        ],
        [
            0.0,
            0.4374999999999999,
        ],
        [
            0.0,
            0.3749999999999999,
        ],
        [
            0.0,
            0.3124999999999999,
        ],
        [
            0.0,
            0.2499999999999999,
        ],
        [
            0.0,
            0.1874999999999999,
        ],
        [
            0.0,
            0.1249999999999999,
        ],
        [
            0.0,
            0.0625,
        ],
        [
            0.05822998290347076,
            0.2605649934272224,
        ],
        [
            0.2690625499678753,
            0.4373741634969621,
        ],
        [
            0.4394063223897706,
            0.2532266282495676,
        ],
        [
            0.218731571670798,
            0.05413444603683523,
        ],
        [
            0.4485296952607269,
            0.3514543919622855,
        ],
        [
            0.1562572613976605,
            0.4430017882804914,
        ],
        [
            0.3139361519575364,
            0.05655977493085765,
        ],
        [
            0.06317620356498166,
            0.143322085838519,
        ],
        [
            0.09487783772320771,
            0.05177176608767001,
        ],
        [
            0.4346802809668467,
            0.1254847795829751,
        ],
        [
            0.385835561124088,
            0.4424107541217095,
        ],
        [
            0.05642710595914754,
            0.3800935995942247,
        ],
        [
            0.2206910771762775,
            0.4223319654557284,
        ],
        [
            0.2409669046111134,
            0.3803342651533306,
        ],
        [
            0.2796523091247737,
            0.3809670135891468,
        ],
        [
            0.2811532273431561,
            0.3278775056845646,
        ],
        [
            0.2226036411773187,
            0.3311968358436785,
        ],
        [
            0.2383195344809018,
            0.2774912466426128,
        ],
        [
            0.2785950269114238,
            0.275688209913682,
        ],
        [
            0.2811213160111221,
            0.22424678930415,
        ],
        [
            0.2051718789543341,
            0.2190264006807288,
        ],
        [
            0.2437323515129852,
            0.1510889946253526,
        ],
        [
            0.2955104382677184,
            0.1705633178693746,
        ],
        [
            0.170775327327602,
            0.1496636539918524,
        ],
        [
            0.1214950554939795,
            0.1549693083706915,
        ],
        [
            0.1571075604841364,
            0.1175566591042925,
        ],
        [
            0.3096864964698186,
            0.1137240137759864,
        ],
        [
            0.3600135006231731,
            0.1960897704484206,
        ],
        [
            0.1223462154153212,
            0.2116835515605153,
        ],
        [
            0.1160915762585152,
            0.2719530266180434,
        ],
        [
            0.1103153596848872,
            0.3288204042469799,
        ],
        [
            0.1503778448299227,
            0.3343952793879381,
        ],
        [
            0.1108085553193199,
            0.3847477195247274,
        ],
        [
            0.1722344345729996,
            0.388665431896197,
        ],
        [
            0.4451386678415112,
            0.3111850878656598,
        ],
        [
            0.3810973352747714,
            0.2611286164327285,
        ],
        [
            0.4349323164015576,
            0.1902737625983149,
        ],
        [
            0.2602460266589819,
            0.05295883471215043,
        ],
        [
            0.169321158311335,
            0.2933206700284435,
        ],
        [
            0.3901133716683218,
            0.3173464068481684,
        ],
        [
            0.4042022912715102,
            0.3789676816396894,
        ],
        [
            0.1591573024428046,
            0.04559816201766287,
        ],
        [
            0.3260818403833355,
            0.4401545419132589,
        ],
        [
            0.2067505282756192,
            0.1228710534088913,
        ],
        [
            0.3737227477880348,
            0.06011343705045663,
        ],
        [
            0.4477468983095898,
            0.402878525370986,
        ],
        [
            0.07366106010639908,
            0.09356062491854583,
        ],
        [
            0.1150073732639813,
            0.4420073499310098,
        ],
        [
            0.3344198772063799,
            0.3243065221373092,
        ],
        [
            0.3216841712974912,
            0.2322442333286768,
        ],
        [
            0.1243251727668242,
            0.1096855974361544,
        ],
        [
            0.3697587371967951,
            0.1225195262768584,
        ],
        [
            0.06043027625340477,
            0.2006739036270033,
        ],
        [
            0.05640733571939122,
            0.3204727192992418,
        ],
        [
            0.05704756158401425,
            0.03578978132756394,
        ],
        [
            0.4363834543890258,
            0.0619771294693556,
        ],
        [
            0.4434497320434642,
            0.443296169676297,
        ],
        [
            0.05864054073426955,
            0.4399127167621574,
        ],
        [
            0.3286863771347082,
            0.2710301016990295,
        ],
        [
            0.2280270786306877,
            0.09529701131120621,
        ],
        [
            0.1773829335529767,
            0.09451185524609806,
        ],
        [
            0.2598706027949543,
            0.1025253630660608,
        ],
        [
            0.1405036776174168,
            0.07236710706135094,
        ],
        [
            0.3345858245375794,
            0.3811534603906345,
        ],
    ],
    connectivity: [
        Quad4d2Connectivity(
            [
                67,
                59,
                68,
                34,
            ],
        ),
        Quad4d2Connectivity(
            [
                76,
                83,
                58,
                38,
            ],
        ),
        Quad4d2Connectivity(
            [
                56,
                60,
                84,
                39,
            ],
        ),
        Quad4d2Connectivity(
            [
                64,
                43,
                85,
                62,
            ],
        ),
        Quad4d2Connectivity(
            [
                13,
                68,
                41,
                12,
            ],
        ),
        Quad4d2Connectivity(
            [
                70,
                48,
                65,
                63,
            ],
        ),
        Quad4d2Connectivity(
            [
                70,
                63,
                62,
                61,
            ],
        ),
        Quad4d2Connectivity(
            [
                78,
                40,
                94,
                82,
            ],
        ),
        Quad4d2Connectivity(
            [
                65,
                64,
                62,
                63,
            ],
        ),
        Quad4d2Connectivity(
            [
                65,
                37,
                79,
                64,
            ],
        ),
        Quad4d2Connectivity(
            [
                44,
                45,
                46,
                33,
            ],
        ),
        Quad4d2Connectivity(
            [
                34,
                14,
                15,
                66,
            ],
        ),
        Quad4d2Connectivity(
            [
                71,
                67,
                34,
                66,
            ],
        ),
        Quad4d2Connectivity(
            [
                72,
                71,
                66,
                36,
            ],
        ),
        Quad4d2Connectivity(
            [
                16,
                36,
                66,
                15,
            ],
        ),
        Quad4d2Connectivity(
            [
                48,
                47,
                46,
                45,
            ],
        ),
        Quad4d2Connectivity(
            [
                44,
                65,
                48,
                45,
            ],
        ),
        Quad4d2Connectivity(
            [
                35,
                73,
                5,
                6,
            ],
        ),
        Quad4d2Connectivity(
            [
                7,
                8,
                38,
                69,
            ],
        ),
        Quad4d2Connectivity(
            [
                68,
                13,
                14,
                34,
            ],
        ),
        Quad4d2Connectivity(
            [
                65,
                44,
                22,
                37,
            ],
        ),
        Quad4d2Connectivity(
            [
                52,
                70,
                61,
                60,
            ],
        ),
        Quad4d2Connectivity(
            [
                42,
                19,
                20,
                74,
            ],
        ),
        Quad4d2Connectivity(
            [
                33,
                21,
                22,
                44,
            ],
        ),
        Quad4d2Connectivity(
            [
                69,
                35,
                6,
                7,
            ],
        ),
        Quad4d2Connectivity(
            [
                48,
                70,
                52,
                49,
            ],
        ),
        Quad4d2Connectivity(
            [
                52,
                53,
                54,
                51,
            ],
        ),
        Quad4d2Connectivity(
            [
                52,
                60,
                56,
                55,
            ],
        ),
        Quad4d2Connectivity(
            [
                48,
                49,
                50,
                47,
            ],
        ),
        Quad4d2Connectivity(
            [
                52,
                51,
                50,
                49,
            ],
        ),
        Quad4d2Connectivity(
            [
                74,
                20,
                21,
                33,
            ],
        ),
        Quad4d2Connectivity(
            [
                52,
                55,
                75,
                53,
            ],
        ),
        Quad4d2Connectivity(
            [
                61,
                32,
                84,
                60,
            ],
        ),
        Quad4d2Connectivity(
            [
                22,
                23,
                79,
                37,
            ],
        ),
        Quad4d2Connectivity(
            [
                61,
                62,
                85,
                32,
            ],
        ),
        Quad4d2Connectivity(
            [
                72,
                36,
                16,
                77,
            ],
        ),
        Quad4d2Connectivity(
            [
                8,
                9,
                76,
                38,
            ],
        ),
        Quad4d2Connectivity(
            [
                31,
                78,
                39,
                30,
            ],
        ),
        Quad4d2Connectivity(
            [
                30,
                39,
                84,
                29,
            ],
        ),
        Quad4d2Connectivity(
            [
                28,
                32,
                85,
                27,
            ],
        ),
        Quad4d2Connectivity(
            [
                54,
                58,
                83,
                59,
            ],
        ),
        Quad4d2Connectivity(
            [
                17,
                2,
                18,
                88,
            ],
        ),
        Quad4d2Connectivity(
            [
                25,
                89,
                24,
                3,
            ],
        ),
        Quad4d2Connectivity(
            [
                4,
                86,
                31,
                0,
            ],
        ),
        Quad4d2Connectivity(
            [
                11,
                87,
                10,
                1,
            ],
        ),
        Quad4d2Connectivity(
            [
                72,
                42,
                74,
                95,
            ],
        ),
        Quad4d2Connectivity(
            [
                59,
                81,
                51,
                54,
            ],
        ),
        Quad4d2Connectivity(
            [
                71,
                72,
                95,
                80,
            ],
        ),
        Quad4d2Connectivity(
            [
                55,
                56,
                82,
                57,
            ],
        ),
        Quad4d2Connectivity(
            [
                47,
                50,
                90,
                80,
            ],
        ),
        Quad4d2Connectivity(
            [
                47,
                80,
                95,
                46,
            ],
        ),
        Quad4d2Connectivity(
            [
                59,
                83,
                41,
                68,
            ],
        ),
        Quad4d2Connectivity(
            [
                56,
                39,
                78,
                82,
            ],
        ),
        Quad4d2Connectivity(
            [
                51,
                81,
                90,
                50,
            ],
        ),
        Quad4d2Connectivity(
            [
                32,
                28,
                29,
                84,
            ],
        ),
        Quad4d2Connectivity(
            [
                16,
                17,
                88,
                77,
            ],
        ),
        Quad4d2Connectivity(
            [
                9,
                10,
                87,
                76,
            ],
        ),
        Quad4d2Connectivity(
            [
                43,
                26,
                27,
                85,
            ],
        ),
        Quad4d2Connectivity(
            [
                19,
                42,
                88,
                18,
            ],
        ),
        Quad4d2Connectivity(
            [
                26,
                43,
                89,
                25,
            ],
        ),
        Quad4d2Connectivity(
            [
                12,
                41,
                87,
                11,
            ],
        ),
        Quad4d2Connectivity(
            [
                5,
                40,
                86,
                4,
            ],
        ),
        Quad4d2Connectivity(
            [
                59,
                67,
                90,
                81,
            ],
        ),
        Quad4d2Connectivity(
            [
                23,
                24,
                89,
                79,
            ],
        ),
        Quad4d2Connectivity(
            [
                93,
                91,
                35,
                69,
            ],
        ),
        Quad4d2Connectivity(
            [
                58,
                93,
                69,
                38,
            ],
        ),
        Quad4d2Connectivity(
            [
                80,
                90,
                67,
                71,
            ],
        ),
        Quad4d2Connectivity(
            [
                75,
                92,
                35,
                91,
            ],
        ),
        Quad4d2Connectivity(
            [
                93,
                53,
                75,
                91,
            ],
        ),
        Quad4d2Connectivity(
            [
                54,
                53,
                93,
                58,
            ],
        ),
        Quad4d2Connectivity(
            [
                55,
                57,
                92,
                75,
            ],
        ),
        Quad4d2Connectivity(
            [
                35,
                92,
                94,
                73,
            ],
        ),
        Quad4d2Connectivity(
            [
                5,
                73,
                94,
                40,
            ],
        ),
        Quad4d2Connectivity(
            [
                33,
                46,
                95,
                74,
            ],
        ),
        Quad4d2Connectivity(
            [
                83,
                76,
                87,
                41,
            ],
        ),
        Quad4d2Connectivity(
            [
                31,
                86,
                40,
                78,
            ],
        ),
        Quad4d2Connectivity(
            [
                72,
                77,
                88,
                42,
            ],
        ),
        Quad4d2Connectivity(
            [
                64,
                79,
                89,
                43,
            ],
        ),
        Quad4d2Connectivity(
            [
                92,
                57,
                82,
                94,
            ],
        ),
    ],
}
